### ISSUE TITLE
Implement Query Source Type

### ISSUE THREAD

**Issue by Vinai** [hyva-themes/magento2-hyva-admin#17]
Allow developers to display db table content directly without having to implement intermediate PHP classes.

Roughly I had something like this in mind (**EDIT: SEE [UPDATED VERSION IN THREAD BELOW](https://github.com/hyva-themes/magento2-hyva-admin/issues/17#issuecomment-814735524)**)):

```xml
<?xml version="1.0"?>
<grid xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:module:Hyva_Admin:etc/hyva-grid.xsd">
    <source>
        <query>
            <columns>
                <column as="id">entity_id</column>
                <column>sku</column>
                <column as="name">t_name.value</column>
            </columns>
            <from as="mainTable">catalog_product</from>
            <where>
                <and>id IN(:ids)</and>
                <or>sku = :sku</or>
            </where>
            <join>
                <left conditionType="with">
                    <table as="t_name">catalog_product_entity_varchar</table>
                    <condition>t_name.entity_id=mainTable.entity_id AND attribute_id=47</condition>
                </left>
            </join>
            <bind_parameters>
                <param name="ids">productIds</param>
                <param name="sku">sku</param>
            </bind_parameters>
        </query>
    </source>
</grid>
```
This is a very rough draft  and unfinished, all open to change. 
It is first attempt to come up with a intuitive XML schema to allow defining a query in an extensible manner.

Most of the above should hopefully be understandable, but the bind parameters might need an explanation:
The use case for the bind params is to define a grid that is (for example) on a product page, and that then displays some data for that product.
There needs to be some way to pass values to the grid from PHP via the bind parameters.

In this example, the grid grid block (?) could say something like
```php
$grid->addSourceBindParams([
    'productIds' => $ids,
    'sku' => $request->getParam('sku')
]);
```

You might notice this is not completely thought out yet.

**Comment by Vinai** [comment]
In Hyva_Admin version 1.1.5 the `\Hyva\Admin\Model\TypeReflection\DbSelectColumnExtractor` was added, which can inspect `\Magento\Framework\DB\Select` instances to figure out columns and types.

Also, `\Hyva\Admin\Model\MethodValueBindings` provides all that is needed to pass values from PHP into the query.

All that is missing is the code to build the `Select` instance from the XML.

Note: the XML draft above needs to be expanded to include table name aliasing.

**Comment by Vinai** [comment]
Reworked draft for the query source type XML:

```xml
<?xml version="1.0"?>
<grid xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:noNamespaceSchemaLocation="urn:magento:module:Hyva_Admin:etc/hyva-grid.xsd">
    <source>
        <query>
           <select>
              <from as="main_table">catalog_product</from>
              <columns>
                  <column as="id">entity_id</column>
                  <column>sku</column>
              </columns>
              <join type="left" table="catalog_product_entity_varchar" as="t_name">
                  <condition>t_name.entity_id=main_table.entity_id AND attribute_id=47</condition>
                  <columns>
                    <column as="name">value</column>
                  </columns>
              </join>
           </select>
           <unionSelect type="all">
              ... all contents of select allowed ...
           </unionSelect>
        </query>
    </source>
</grid>
```

Nested subqueries will not be supported in the initial version.
Any conditions are specified using the `<defaultSearchCriteriaBindings>` outside of the `<query>` spec.

If no columns are specified on the main select, `*` is used.
For joined tables it is different: if no columns are specified on a join, no columns are added to the result set. It is possible to specify `*` as a join column, too.

**Comment by Vinai** [comment]
Implement in future if it turns out they are needed:

- having
- subqueries

**Comment by Vinai** [comment]
Feature complete in fb21882, now moving into manual testing and documentation.

**Comment by Vinai** [comment]
Implemented in eaa9072

Still missing: documentation and the features for future enhancements listed in https://github.com/hyva-themes/magento2-hyva-admin/issues/17#issuecomment-814982844

**Comment by Vinai** [comment]
First query source type released in 1.1.6.
Closing this issue for now. If one of the missing features above needs to be implemented, I'll reopen.
